### PR TITLE
feat: new internal macro for rate-limited interruption checks

### PR DESCRIPTION
**Pull request by szhorvat**
This is a suggestion to add a macro that allows limiting the number of interruption checks, and thus increasing performance.

### CODE REVIEW

File: src/constructors/regular.c [hunk 1]
```diff
@@ -410,7 +411,7 @@ igraph_error_t igraph_square_lattice(
 #define IS_PERIODIC(dim) ((periodic && VECTOR(*periodic)[dim]))
 
     for (i = 0; i < no_of_nodes; i++) {
-        IGRAPH_ALLOW_INTERRUPTION();
+        IGRAPH_ALLOW_INTERRUPTION_LIMITED(iter, 1 << 10);
```
Comment (ntamas) [inline src/constructors/regular.c:414]: Can't we use `i` here directly?

File: src/core/interruption.h [hunk 2]
```diff
@@ -52,6 +52,21 @@ extern IGRAPH_THREAD_LOCAL igraph_interruption_handler_t *igraph_i_interruption_
         } \
     } while (0)
 
+/**
+ * \define IGRAPH_ALLOW_INTERRUPTION_LIMITED
+ *
+ * This is a variant of IGRAPH_ALLOW_INTERRUPTION() that checks for interruption
+ * only on every 'skips' call. It uses the 'iter' variable, which must be declared
+ * separately and initialized to 0, to keep track of the number of iterations.
```
Comment (ntamas) [inline src/core/interruption.h:60]: If we enforce the iteration tracker variable to be named as `iter` then there is no need to pass it as an argument to the macro. I'd suggest calling it `__iter` in the _macro_ and keeping the first argument of the macro so the user can choose whichever variable they want in the code to be used as an iteration variable.

File: src/paths/simple_paths.c [hunk 3]
```diff
@@ -152,10 +148,7 @@ igraph_error_t igraph_get_all_simple_paths(const igraph_t *graph,
             VECTOR(nptr)[up] = 0;
         }
 
-        iteration++;
-        if (iteration >= 10000) {
-            iteration = 0;
-        }
+        IGRAPH_ALLOW_INTERRUPTION_LIMITED(iter, 1 << 13);
```
Comment (ntamas) [inline src/paths/simple_paths.c:151]: I guess `1 << 13` is okay but the original code used `10000` so I wonder whether this is a typo.

File: src/constructors/regular.c (hunk 1 above)
Comment (szhorvat) [reply src/constructors/regular.c:414]: If we do that, we need to use `i % skips == 0`. As I recall, `%` was slow, or perhaps not as amenable to compiler optimization across the loop. But I did not test this now.

File: src/core/interruption.h (hunk 2 above)
Comment (szhorvat) [reply src/core/interruption.h:60]: Sorry, I forgot the update the comment after I made the variable name a parameter.

File: src/constructors/regular.c (hunk 1 above)
Comment (ntamas) [reply src/constructors/regular.c:414]: Good point!

File: src/core/interruption.h (hunk 2 above)
Comment (ntamas) [reply src/core/interruption.h:60]: Okay, no worries. If you fix the comments, then we're done, feel free to merge then. Let me know when you think it's time for a new C patch release.

**Review by ntamas — APPROVED** [review]

File: src/core/interruption.h (hunk 2 above)
Comment (szhorvat) [reply src/core/interruption.h:60]: @ntamas See chat about C patch release.

**Pull request merged**